Prevent handle leaks and double-frees

uckelman-sf opened the pull request:
This PR wraps almost all local uses of TSK handles in `std::unique_ptr`, to ensure that they are freed exactly once along every code path.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Removed duplicate code.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Don't leak path.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Ensure that TSK_FS_FILE::name and meta are initialized.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

simsong commented:
Hi @uckelman-sf  — thank you so much for your attention to this! Please let me know if there is anything I can do. Please feel free to update the CI/CD system if you think we can get more info about if it.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Use std::unique_ptr for handles to prevent leaks and double-frees.

Commit:
Fixed literal-suffix warning.

Commit:
Fix sign-compare warning: (dentry->flags & 0x01) is not a uint8_t---it gets
promoted---so when you bitwise negate it, you end up turning on some bits
that didn't exist in dentry->flags, which means that ~(dentry->flags & 0x01)
is always true. That cannot be what was intended as the condition here;
the comment suggests that logical negation is wanted instead of bitwise,
i.e., the test is for whether the low bit is zero.

simsong committed:
Merge branch 'develop' into prevent_handle_leaks

simsong commented:
Well, this all works except for the appveyor, which I need to fix, so I'm going to merge this in.

uckelman commented:
You're probably missing `SLEUTKHIT_TEST_DATA_DIR` in the appveyor build as well.

simsong commented:
> You're probably missing `SLEUTKHIT_TEST_DATA_DIR` in the appveyor build as well.

Indeed, but I think that I added it. See: https://github.com/sleuthkit/sleuthkit/pull/3258

I'm now working on the CI/CD thing for the most recent SleuthKit release from CyberTriage which demonstrates that the CT fork also has the problem that you just fixed...
